### ISSUE TITLE
TCP/UDP

### ISSUE THREAD

**Issue by aleczdr** [IPCams/BugTracker#12]
This is kinda both a bug and a feature request
**Is your feature request related to a problem? Please describe.**
In versions of the app prior to the ui upgrade, i could view my cameras from mobile data. Since the upgrade, the stream fails to start. They still work when im on my home wifi.

**Describe the solution you'd like**
I think a switch that chooses tcp/udp for rtsp would fix the problem 

**Additional context**
I had a similar problem with ffmpeg/ffplay on my computer which i was able to fix by disabling tcp.
There was no issue prior to the ui upgrade.
This may have something to do with my port fowarding as i can’t change the port on the camera so my router is mapping it to a different external port, but it was working before.

**Comment by fmarcoux96** [comment]
Yes, this is needed. Specially when using RTSP. It takes forever to start the feed in UDP via VPN.

**Comment by tylerjonesio** [comment]
This is something I have been looking into. By default the stream type should default to UDP if available and fall back to TCP if not. I'm not sure if I can add a force UDP only option since that is the default, but I do have an option to force a TCP stream or force the stream to tunnel over HTTP. 

Since this is not something I can easily test out with my cameras, I have put these options behind an `Experimental Mode` until I can verify this works as expected. To enable experimental mode, go to the settings tab and tap the app icon at the bottom five times. This should then show a message and give haptic feedback when enabled.

Once experimental mode has been enabled, you can then edit your camera and specify the stream type under the advanced section. Auto is the default behavior which prefers UDP with a TCP fallback, TCP forces a TCP stream, and HTTP forces a stream via HTTP tunneling. 

Feel free to try this out and let me know how and if it works for you all. If it works as expected I can eventually push this out to everyone.

|     |     |
|---|---|
|![Simulator Screen Shot - iPhone 11 Pro - 2020-10-17 at 09 37 26](https://user-images.githubusercontent.com/8380082/96347198-5d678380-105d-11eb-858e-50ef21677c25.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2020-10-17 at 09 37 09](https://user-images.githubusercontent.com/8380082/96347204-65bfbe80-105d-11eb-8e58-17659d44e821.png) |

**Comment by fmarcoux96** [comment]
@tylerjonesio I tried the TCP option! Much faster locally, and over VPN as well! Really cool!

Are there more experiments on the way? 😄

**Comment by tylerjonesio** [comment]
Awesome! I'm glad to hear that it is working for you! Now that I have confirmation that it works, I'll look at getting this pushed out to all. 

I don't have any other experiments planned at the moment, but I'm sure there will be in the future. I figured I needed a way to get extra testing on certain things before a wide release, thus experimental mode was born.

**Comment by kezzawd** [comment]
Fantastic - have just downloaded and was wondering how to send my rtsp streams over tcp and voila - 5 taps does it! Thank you for adding this option - going Pro with confidence now 😁

**Comment by aleczdr** [comment]
i can confirm- tcp makes it much faster. unfortunately it still doesn't work over mobile data. it wont load on both lte and 3g.

**Comment by fmarcoux96** [comment]
@aleczdr Is your IP even accessible from the outside? With NAT and everything? Because TCP won't make a closed port opened...

**Comment by aleczdr** [comment]
sorry for the late reply
@fmarcoux96 yes it is accessible from outside, it worked before when i had an older version of the app. It had stopped working right around the time of the app's ui upgrade. I had ios 12 then so i know it wasn't a "allow access to local network" ios privacy setting or something similar. i tried disabling my router's firewall and it still didn't work. I've been meaning to try it on a different wifi network but due to current events thats not really possible for the moment. If i was able v1.6 i could test to make sure that its not an issue with verizon or ios (although it probably isnt), or my network settings.

**Comment by fmarcoux96** [comment]
> sorry for the late reply
> @fmarcoux96 yes it is accessible from outside, it worked before when i had an older version of the app. It had stopped working right around the time of the app's ui upgrade. I had ios 12 then so i know it wasn't a "allow access to local network" ios privacy setting or something similar. i tried disabling my router's firewall and it still didn't work. I've been meaning to try it on a different wifi network but due to current events thats not really possible for the moment. If i was able to obtain app version 1.6 i could test to make sure that its not an issue with verizon or ios (although it probably isnt), or my network settings.

If you're on iOS 12, that is why. Not supported anymore.

**Comment by aleczdr** [comment]
> > sorry for the late reply
> > @fmarcoux96 yes it is accessible from outside, it worked before when i had an older version of the app. It had stopped working right around the time of the app's ui upgrade. I had ios 12 then so i know it wasn't a "allow access to local network" ios privacy setting or something similar. i tried disabling my router's firewall and it still didn't work. I've been meaning to try it on a different wifi network but due to current events thats not really possible for the moment. If i was able to obtain app version 1.6 i could test to make sure that its not an issue with verizon or ios (although it probably isnt), or my network settings.
> 
> If you're on iOS 12, that is why. Not supported anymore.

i had ios 12 when i updated to v2.0. thats when it stopped working. then i updated to ios 14 and it still didnt work. i've had ios 14 since september

**Comment by fmarcoux96** [comment]
> > > sorry for the late reply
> > > @fmarcoux96 yes it is accessible from outside, it worked before when i had an older version of the app. It had stopped working right around the time of the app's ui upgrade. I had ios 12 then so i know it wasn't a "allow access to local network" ios privacy setting or something similar. i tried disabling my router's firewall and it still didn't work. I've been meaning to try it on a different wifi network but due to current events thats not really possible for the moment. If i was able to obtain app version 1.6 i could test to make sure that its not an issue with verizon or ios (although it probably isnt), or my network settings.
> > 
> > 
> > If you're on iOS 12, that is why. Not supported anymore.
> 
> i had ios 12 when i updated to v2.0. thats when it stopped working. then i updated to ios 14 and it still didnt work. i've had ios 14 since september

My guess is your carrier somehow might block the UDP/TCP traffic on RTSP.  It works flawlessly for me.

Have you tried setting up a VPN between your iPhone and your router? This way, you enable VPN to view your cameras, but bypasses any carrier/provider restrictions.

**Comment by aleczdr** [comment]
> > > > sorry for the late reply
> > > > @fmarcoux96 yes it is accessible from outside, it worked before when i had an older version of the app. It had stopped working right around the time of the app's ui upgrade. I had ios 12 then so i know it wasn't a "allow access to local network" ios privacy setting or something similar. i tried disabling my router's firewall and it still didn't work. I've been meaning to try it on a different wifi network but due to current events thats not really possible for the moment. If i was able to obtain app version 1.6 i could test to make sure that its not an issue with verizon or ios (although it probably isnt), or my network settings.
> > > 
> > > 
> > > If you're on iOS 12, that is why. Not supported anymore.
> > 
> > 
> > i had ios 12 when i updated to v2.0. thats when it stopped working. then i updated to ios 14 and it still didnt work. i've had ios 14 since september
> 
> My guess is your carrier somehow might block the UDP/TCP traffic on RTSP. It works flawlessly for me.
> 
> Have you tried setting up a VPN between your iPhone and your router? This way, you enable VPN to view your cameras, but bypasses any carrier/provider restrictions.

yes vpn works. i dont think frontier (verizon subcontractor) is blocking rtsp. it wouldnt get blocked at the exact same moment i updated the app.

**Comment by tylerjonesio** [comment]
I don't know if you are still having this issue, but I did want to let you know that I am still looking into this. The current underlying video streaming libraries that are used don't give much control other than what's already surfaced up. And because of that, I'm looking into what I can do to eventually replace it with a more modern ffmpeg backed player. 

As mentioned [here](https://github.com/IPCams/BugTracker/issues/43#issuecomment-820009560), version 2.3 has a new experimental video player that I'm working on. It's not 100% ready and doesn't cover all of the existing RTSP use cases yet, but it does provide a much better foundation. Feel free to try it out if you'd like and let me know how it works for you. That would be great if it does work, but if it doesn't there is more control that I have to hopefully get to a working solution.

**Comment by aleczdr** [comment]
I just tried the new video player-- it doesn't work at all. Not even on wifi. The old one still works fine on wifi.

Speaking of ffmpeg, when i use it on my mac to view the camera (i have v4.2), i have to use different addresses depending if im on the same network as the camera or if im connecting from the outside. This may be an issue if it is implemented in the app. 

However, in the current ipcams app viewer i can use the same address i use for the external connection on my mac regardless if i am connected to my home wifi, but, since i upgraded from ipcams v1.6 i have not been able to get the camera to load unless im on my home wifi.

I have gotten a new wifi router since this issue originally occurred, and while it has made no difference when using ffmpeg/ffplay on my mac, having access to app v1.6 would allow me to see test if this is indeed an app issue or if its a wifi network issue, or, since verizon has pretty much rebuilt the entire cell tower network in my area (winter storm damage and the 5g rollout), a verizon issue. Or, if it is possible to implement the player from ipcams v1.6 into the current version, i could switch the router back and test the camera with that version to see if that still works. 

On a side note, the TCP function has worked very well, and has drastically reduced load times.

**Comment by tylerjonesio** [comment]
Thanks for testing that out. Even though it didn't work, it is great feedback! Do you happen to know if your cameras are using an h265 stream? I've identified a compatibility issue with h265 streams in the experimental video player that I've hopefully fixed in 2.3.3 which is currently rolling out.

If that doesn't happen to fix things, would you mind letting me know what ffmpeg/ffplay commands you are using (with redacted host/auth info)? I'd love to get your use case (and any others that I might be missing) accounted for. Once the new player is more solid and battle tested, I'll eventually release it to everyone. That will unblock and unlock some other issues/features (like PIP) that I can't just support with the current player.

In terms of being able to test the currently player in older versions of the app, there's really not much I can do there. It was using a previous version of the libraries currently used for rtsp cams and there is no way to roll that back without causing issues for others. Not sure what the difference would have been between versions as it was just minor version bumps. I am curious though if there might be something more than just TCP vs UDP at play. Maybe something like IPV6 issues on cellular as reported in #43? Whatever the issue is, if there is a way to fix what is happening in IPCams for you, it'll probably have to happen via the new video player. Mainly because I have a lot more control over it vs the old one.

I did want to highlight your comment about needing to use different addresses if you are internal vs external. If that becomes an issue, in the short term you could just duplicate your existing camera config within IPCams and then just change the host/ip info. I currently do that myself for some of my cameras. I know it isn't the most optimal solution though. Long term if it is of any interest, I could look into adding an option for an external host/address. Not sure how difficult that would be, but I have seen other apps that do similar so it should be possible.

**Comment by aleczdr** [comment]
My cameras can do h264, h265, and analog. I currently have them set to h265 but i'll try one on h264 later and see if that works. As far as i know 3G is ipv4 only, and frontier/verizon fiber doesn't provide ipv6 service, so even on lte i'd still be using ipv4.
i use the following ffplay command:
`ffplay -i rtsp://`user:pw@host:port`/h265major -window_title "Camera 1" -stats -vf scale=720x480,setdar=16/9`
where the hostname and port is different depending if internal/external. I do use udp in ffplay because it loads faster than tcp, which is the complete opposite of the app.

**Comment by aleczdr** [comment]
Alright, so it turns out the problem **was** verizon. go figure.

Verizon replaced my ont yesterday. After the swap the cameras magically worked. I don't know why it only affected rtsp streams (as far as i could tell), but now the cameras work on wifi, 3g, and lte. So it wasn't the app.

**Comment by fmarcoux96** [comment]
> Alright, so it turns out the problem **was** verizon. go figure.
> 
> Verizon replaced my ont yesterday. After the swap the cameras magically worked. I don't know why it only affected rtsp streams (as far as i could tell), but now the cameras work on wifi, 3g, and lte. So it wasn't the app.

That is indeed weird as f***! An ONT should have no effect on a stream but, okay I guess? Glad to hear lol
